src/Microsoft.AspNet.Server.Kestrel/Http/ListenerSecondary.cs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNet.Server.Kestrel.Networking;
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNet.Server.Kestrel.Http
+{
+    public class ListenerSecondary : ListenerContext, IDisposable
+    {
+        UvPipeHandle DispatchPipe { get; set; }
+
+        public ListenerSecondary(IMemoryPool memory)
+        {
+            Memory = memory;
+        }
+
+        public Task StartAsync(
+            string pipeName,
+            KestrelThread thread,
+            Func<Frame, Task> application)
+        {
+            Thread = thread;
+            Application = application;
+
+            DispatchPipe = new UvPipeHandle();
+
+            var tcs = new TaskCompletionSource<int>();
+            Thread.Post(_ =>
+            {
+                try
+                {
+                    DispatchPipe.Init(Thread.Loop, true);
+                    var connect = new UvConnectRequest();
+                    connect.Init(Thread.Loop);
+                    connect.Connect(
+                        DispatchPipe,
+                        pipeName,
+                        (connect2, status, error, state) =>
+                        {
+                            connect.Dispose();
+                            if (error != null)
+                            {
+                                tcs.SetException(error);
+                                return;
+                            }
+
+                            try
+                            {
+                                var ptr = Marshal.AllocHGlobal(16);
+                                var buf = Thread.Loop.Libuv.buf_init(ptr, 16);
+
+                                DispatchPipe.ReadStart(
+                                    (_1, _2, _3) => buf,
+                                    (_1, status2, error2, state2) =>
+                                    {
+                                        if (status2 <= 0)
```

Comment:
Right now, the stream's read implementation will only ever give 0 (for any kind of error) or >0 for successful reads